### MovieLens Case Study

The GroupLens Research Project is a research group in the Department of Computer Science and Engineering at the University of Minnesota. The data is widely used for collaborative filtering and other filtering solutions. However, we will be using this data to act as a means to demonstrate our skill in using Python to “play” with data.

### Datasets Information:

- **ratings.csv:** It contains information on ratings given by the users to a particular movie. Columns: user id, movie id, rating, timestamp

- **movie.csv:** The file contains information related to the movies and their genre. Columns: movie id, movie title, release date, unknown, Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western

- **user.csv:** It contains information about the users who have rated the movies.. Columns: user id, age, gender, occupation, zip code

### Objective:

`To extract insights from the dataset`

### Learning Outcomes:
`Use of Pandas Functions - shape, describe, groupby, merge etc.`


#### Domain 
`Internet and Entertainment`

**Note that the case study will need you to apply the concepts of groupby and merging extensively.**

In [44]:
# Kaspar Edit:  my favorite extension for Notebooks
!conda install nb_black

^C
Solving environment: ...working... done

# All requested packages already installed.



In [45]:
%load_ext lab_black

The lab_black extension is already loaded. To reload it, use:
  %reload_ext lab_black


In [47]:
## Research & read all of PEP8


def messy_func(
    x=50,
    long_variable_name_that_takes_up_space="hjah34j5k2h345kjh4",
    another_long_variable="ap2934y5hlkjahlrkjaherkljhaeljkrhalkjsdfn",
):
    return str(x) + long_variable_name_that_takes_up_space

### 1. Import the necessary packages

In [48]:
import pandas as pd
import numpy as np

### 2. Read all the three datasets

In [50]:
# Reading datasets by using read_csv from pandas package
ratings = pd.read_csv("ratings.csv")
movie = pd.read_csv("movie.csv")
user = pd.read_csv("user.csv")

### 3. View the first 5 rows of all the datasets.
`Note that you will need to do it for all three datasets separately`

In [51]:
ratings.head(5)

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [55]:
movie.head(5).T

0           1            2            3         4
movie id               1           2            3            4         5
movie title   Toy Story   GoldenEye   Four Rooms   Get Shorty   Copycat 
release date    1-Jan-95    1-Jan-95     1-Jan-95     1-Jan-95  1-Jan-95
Action                 0           1            0            1         0
Adventure              0           1            0            0         0
Animation              1           0            0            0         0
Childrens              1           0            0            0         0
Comedy                 1           0            0            1         0
Crime                  0           0            0            0         1
Documentary            0           0            0            0         0
Drama                  0           0            0            1         1
Fantasy                0           0            0            0         0
Film-Noir              0           0            0            0         0
Horror                 0           0            0            0         0
Musical                0           0            0            0         0
Mystery                0           0            0            0         0
Romance                0           0            0            0         0
Sci-Fi                 0           0            0            0         0
Thriller               0           1            1            0         1
War                    0           0            0            0         0
Western                0           0            0            0         0

In [56]:
user.head(5)

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

### 4. Understand the shape of all the datasets.
`Note that you will need to do it for all the three datasets seperately`

In [58]:
# ratings
ratings.shape

(100000, 4)

 **Observation:** There are 100000 rows and 4 columns in the ratings dataset

In [59]:
# user
user.shape

(943, 5)

 **Observation:** There are 943 rows and 5 columns in the user dataset

In [60]:
# movie
movie.shape

(1680, 21)

 **Observation:** There are 1680 rows and 21 columns in the movie dataset

### 5. Check the data types of the columns for all the datasets.
 `Note that you will need to do it for all three datasets separately`

In [64]:
for df in [user, movie, ratings]:
    # print(df.__name__)
    display(df.shape)
    display(df.head(5))
    display(ratings.dtypes)
    print("=" * 50)
    print("=" * 50)

(943, 5)

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

(1680, 21)

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

(100000, 4)

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

In [65]:
# ratings
# We use dataframe.dtypes to get the data types of each column
ratings.dtypes

user id      int64
movie id     int64
rating       int64
timestamp    int64
dtype: object

 **Observation:** All columns have integer data type 

In [66]:
# user
user.dtypes

user id        int64
age            int64
gender        object
occupation    object
zip code      object
dtype: object

In [78]:
user["zip code"].nunique()

795

In [77]:
user["zip code"].str[:3].nunique()

382

 **Observations:**
 1. user id and age columns are of integer data types 
 2. gender, occupation and zip code columns are of string data type

In [79]:
# movie
movie.dtypes

movie id         int64
movie title     object
release date    object
Action           int64
Adventure        int64
Animation        int64
Childrens        int64
Comedy           int64
Crime            int64
Documentary      int64
Drama            int64
Fantasy          int64
Film-Noir        int64
Horror           int64
Musical          int64
Mystery          int64
Romance          int64
Sci-Fi           int64
Thriller         int64
War              int64
Western          int64
dtype: object

**Observation:**
1. movie title and release date are of string data type
2. movie id and all genres are of interger data type

### 6. Give a statistical summary for all the datasets.
`Note that you will need to do it for all three datasets separately`

In [82]:
ratings

user id  movie id  rating  timestamp
0          196       242       3  881250949
1          186       302       3  891717742
2           22       377       1  878887116
3          244        51       2  880606923
4          166       346       1  886397596
...        ...       ...     ...        ...
99995      880       476       3  880175444
99996      716       204       5  879795543
99997      276      1090       1  874795795
99998       13       225       2  882399156
99999       12       203       3  879959583

[100000 rows x 4 columns]

In [81]:
# ratings
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

**Observation:** The mean and Median of user ratings are 3.53 & 4.00 respectively

In [84]:
user

user id  age gender     occupation zip code
0          1   24      M     technician    85711
1          2   53      F          other    94043
2          3   23      M         writer    32067
3          4   24      M     technician    43537
4          5   33      F          other    15213
..       ...  ...    ...            ...      ...
938      939   26      F        student    33319
939      940   32      M  administrator    02215
940      941   20      M        student    97229
941      942   48      F      librarian    78209
942      943   22      M        student    77841

[943 rows x 5 columns]

In [83]:
# user
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

**Observation:** The average age of all the users is 34 years while the range lies between 7 to 73 years.

In [14]:
# movie
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

**Observation:** The genres should be in categorical format and not in the numeric because it is of binary class

### The statistical summary of the data will give us an overview of the data by providing key statistics of the different columns in the data.

### 7. Find the number of movies per genre using the movie data

In [86]:
# Getting all the column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [88]:
# Taking all the genre columns and finding the sum for every column
movie[
    [
        "Action",
        "Adventure",
        "Animation",
        "Childrens",
        "Comedy",
        "Crime",
        "Documentary",
        "Drama",
        "Fantasy",
        "Film-Noir",
        "Horror",
        "Musical",
        "Mystery",
        "Romance",
        "Sci-Fi",
        "Thriller",
        "War",
        "Western",
    ]
].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [89]:
# Alternatively, we can also loc function
movie.loc[:, "Action":"Western"].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [90]:
# Sorting the movies across genres
number = movie.loc[:, "Action":"Western"].sum()
number.sort_values(ascending=False)

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

**Observations:**
1. Drama and Comedy are the most common movie genre.
2. Clearly, there are some movies that have more than one genre.

### Applying functions on specific columns of the data will help us in gaining specific insights about the data.

### 8. Find the movies that have more than one genre
`Hint: use sum on the axis = 1`

In [91]:
# Checking column names
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [94]:
# we create a new dataframe using two columns of the movie dataframe
new_movie = movie[["movie id", "movie title"]].copy()
new_movie

movie id                          movie title
0            1                           Toy Story 
1            2                           GoldenEye 
2            3                          Four Rooms 
3            4                          Get Shorty 
4            5                             Copycat 
...        ...                                  ...
1675      1678                          Mat' i syn 
1676      1679                           B. Monkey 
1677      1680                       Sliding Doors 
1678      1681                        You So Crazy 
1679      1682  Scream of Stone (Schrei aus Stein) 

[1680 rows x 2 columns]

In [102]:
new_movie["Number of Genres_bad"] = movie.loc[:, "Action":"Western"].sum(axis=1)
new_movie

movie id                          movie title  Number of Genres  \
0            1                           Toy Story                  3   
1            2                           GoldenEye                  3   
2            3                          Four Rooms                  1   
3            4                          Get Shorty                  3   
4            5                             Copycat                  3   
...        ...                                  ...               ...   
1675      1678                          Mat' i syn                  1   
1676      1679                           B. Monkey                  2   
1677      1680                       Sliding Doors                  2   
1678      1681                        You So Crazy                  1   
1679      1682  Scream of Stone (Schrei aus Stein)                  1   

      Number of Genres_bad  
0                        3  
1                        3  
2                        1  
3                        3  
4                        3  
...                    ...  
1675                     1  
1676                     2  
1677                     2  
1678                     1  
1679                     1  

[1680 rows x 4 columns]

In [103]:
new_movie.columns

Index(['movie id', 'movie title', 'Number of Genres', 'Number of Genres_bad'], dtype='object')

In [104]:
new_movie = new_movie.drop("Number of Genres_bad", axis=1)
new_movie

movie id                          movie title  Number of Genres
0            1                           Toy Story                  3
1            2                           GoldenEye                  3
2            3                          Four Rooms                  1
3            4                          Get Shorty                  3
4            5                             Copycat                  3
...        ...                                  ...               ...
1675      1678                          Mat' i syn                  1
1676      1679                           B. Monkey                  2
1677      1680                       Sliding Doors                  2
1678      1681                        You So Crazy                  1
1679      1682  Scream of Stone (Schrei aus Stein)                  1

[1680 rows x 3 columns]

In [111]:
# Filtering movies that have more than 1 genres
more_than_one = new_movie["Number of Genres"] > 1
has_y = new_movie["movie title"].str.lower().str.contains("y")

new_movie[more_than_one & has_y]

movie id                           movie title  Number of Genres
0            1                            Toy Story                  3
1            2                            GoldenEye                  3
3            4                           Get Shorty                  3
4            5                              Copycat                  3
6            7                       Twelve Monkeys                  2
...        ...                                   ...               ...
1610      1613                           Tokyo Fist                  2
1615      1618                     King of New York                  2
1623      1626  Nobody Loves Me (Keiner liebt mich)                  2
1653      1656                          Little City                  2
1676      1679                            B. Monkey                  2

[197 rows x 3 columns]

**Observation:** 849 movies have more than one genre.

### Creating new columns using existing columns in the dataframe requires domain knowledge, and can help us to understand our data better.

### 9. Find the top 25 movies according to average ratings such that each movie has number of ratings more than 100

Hint : 

1. First find the movies that have more than 100 ratings(use groupby and count). Extract the movie id in a list.
2. Find the average rating of all the movies and sort them in the descending order. 
3. Use isin(list obtained from 1) to filter out the movies which have more than 100 ratings.
4. You will have to use the .merge() function to get the movie titles.

Note: This question will need you to research about groupby and apply your findings. You can find more on groupby on https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html.

In [129]:
ratings["rating"].mean()

3.52986

In [132]:
top_100_by_count_of_ratings = (
    ratings.groupby("movie id")["rating"]
    .count()
    .reset_index()
    .rename({"rating": "num_of_ratings"}, axis=1)
    .sort_values("num_of_ratings", ascending=False)
    .head(100)["movie id"]
    .to_list()
)
top_100_by_count_of_ratings

[50,
 258,
 100,
 181,
 294,
 286,
 288,
 1,
 300,
 121,
 174,
 127,
 56,
 7,
 98,
 237,
 117,
 172,
 222,
 313,
 204,
 405,
 79,
 210,
 151,
 173,
 69,
 748,
 168,
 269,
 257,
 195,
 423,
 9,
 276,
 318,
 22,
 302,
 328,
 96,
 25,
 118,
 15,
 183,
 216,
 176,
 64,
 202,
 234,
 191,
 28,
 89,
 111,
 275,
 12,
 742,
 357,
 82,
 135,
 289,
 238,
 97,
 268,
 546,
 70,
 333,
 186,
 196,
 475,
 153,
 132,
 228,
 125,
 144,
 483,
 194,
 323,
 245,
 197,
 185,
 11,
 282,
 496,
 568,
 301,
 265,
 655,
 182,
 273,
 143,
 180,
 179,
 471,
 161,
 71,
 95,
 8,
 427,
 678,
 322]

In [136]:
less_than_100 = ~ratings["movie id"].isin(top_100_by_count_of_ratings)

hidden_gems = (
    ratings.sort_values(by="rating", ascending=False).loc[less_than_100, :].head(100)
)
hidden_gems

user id  movie id  rating  timestamp
28426      286       707       5  877531975
38655       94       518       5  891720950
16022      108        10       5  879879834
16021      331        81       5  877196702
91519      617       192       5  883788900
...        ...       ...     ...        ...
79381      872       895       5  888478882
75135      788       684       5  880868401
75134      642      1078       5  885604239
46542       90       482       5  891383204
15951      313       498       5  891015144

[100 rows x 4 columns]

In [142]:
hidden_gems.merge(movie[["movie id", "movie title"]], how="inner", on="movie id")

user id  movie id  rating  timestamp            movie title
0       286       707       5  877531975       Enchanted April 
1       123       707       5  879809943       Enchanted April 
2        94       518       5  891720950     Miller's Crossing 
3       108        10       5  879879834           Richard III 
4       331        81       5  877196702  Hudsucker Proxy, The 
..      ...       ...     ...        ...                    ...
95      881       528       5  876538536   Killing Fields, The 
96      872       895       5  888478882              Scream 2 
97      788       684       5  880868401   In the Line of Fire 
98      642      1078       5  885604239      Oliver & Company 
99       90       482       5  891383204      Some Like It Hot 

[100 rows x 5 columns]

In [23]:
# Merging ratings dataset with movie dataset
df_merge = movie.merge(ratings, on = 'movie id', how = 'inner')
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Musical  Mystery  Romance  Sci-Fi  \
0       1      0            0  ...        0        0        0       0   
1       1      0            0  ...        0        0        0       0   
2       1      0            0  ...        0        0        0       0   
3       1      0            0  ...        0        0        0       0   
4       1      0            0  ...        0        0        0       0   

   Thriller  War  Western  user id  rating  timestamp  
0         0    0        0      308       4  887736532  
1         0    0        0      287       5  875334088  
2         0    0        0      148       4  877019411  
3         0    0        0      280       4  891700426  
4         0    0        0       66       3  883601324  

[5 rows x 24 columns]

In [143]:
# Checking the dimensions of the merged dataframe
df_merge.shape

(99990, 24)

In [144]:
# Finding the count of ratings for each movie using groupby() and count()
# reset_index() is used to shift movie title from being the dataframe’s (movie_count’s) index to
# being just a normal column
movie_count = df_merge.groupby(["movie title"])["rating"].count().reset_index()
movie_count.head()

movie title  rating
0  'Til There Was You        9
1               1-900        5
2      101 Dalmatians      109
3        12 Angry Men      125
4                  187      41

In [145]:
# Extracting the movie titles that have more than 100 ratings
movie_100 = movie_count[movie_count["rating"] > 100]["movie title"]
movie_100.head()

2            101 Dalmatians 
3              12 Angry Men 
7     2001: A Space Odyssey 
15           Absolute Power 
16               Abyss, The 
Name: movie title, dtype: object

In [146]:
# Finding average ratings for each movie and sorting them out in descending order
# using groupby() and sort_values() on merged data frame
avg_rating = (
    df_merge.groupby(["movie title"])["rating"]
    .mean()
    .sort_values(ascending=False)
    .reset_index()
)
avg_rating

movie title  rating
0                         Great Day in Harlem, A      5.0
1                                    Prefontaine      5.0
2                         Someone Else's America      5.0
3     Entertaining Angels: The Dorothy Day Story      5.0
4            Marlene Dietrich: Shadow and Light (     5.0
...                                           ...     ...
1652                                   Babyfever      1.0
1653                              Lashou shentan      1.0
1654                            Shadows (Cienie)      1.0
1655       Shadow of Angels (Schatten der Engel)      1.0
1656                                    Power 98      1.0

[1657 rows x 2 columns]

In [149]:
# Extracting movie titles that have more than 100 ratings using movie titles in movie_100 and isin() function
# Displaying top 25 rows only
cond_1 = avg_rating["movie title"].isin(movie_100)
avg_rating[cond_1].head(25)

movie title    rating
15                                    Close Shave, A   4.491071
16                                  Schindler's List   4.466443
17                               Wrong Trousers, The   4.466102
18                                        Casablanca   4.456790
20                         Shawshank Redemption, The   4.445230
21                                       Rear Window   4.387560
22                               Usual Suspects, The   4.385768
23                                         Star Wars   4.358491
24                                      12 Angry Men   4.344000
28                                      Citizen Kane   4.292929
30                             To Kill a Mockingbird   4.292237
31                   One Flew Over the Cuckoo's Nest   4.291667
32                         Silence of the Lambs, The   4.289744
33                                North by Northwest   4.284916
34                                    Godfather, The   4.283293
35                                    Secrets & Lies   4.265432
36                                 Good Will Hunting   4.262626
37                         Manchurian Candidate, The   4.259542
38  Dr. Strangelove or: How I Learned to Stop Worr...  4.252577
39                           Raiders of the Lost Ark   4.252381
40                                           Vertigo   4.251397
44                                           Titanic   4.245714
45                                Lawrence of Arabia   4.231214
47                               Maltese Falcon, The   4.210145
48                          Empire Strikes Back, The   4.204360

### We have just learned how to use groupby() to  group the various categories in the data, and how to use sort_values() to sort the data.

### 10. Find the relationship between user demographics and movie ratings

#### 10.1 Find the mean rating for every occupation 

1. Merge the user dataset with movie and ratings(already merged : df_merge) dataset
2. Apply groupby on occupation

In [150]:
# Merging user dataset with movie and ratings(already merged : df_merge) dataset
df_merge_all = df_merge.merge(user, on="user id", how="inner")

In [152]:
# Group by occupation and aggregate with mean
df_merge_all.groupby("occupation")["rating"].mean()

occupation
administrator    3.635646
artist           3.653380
doctor           3.688889
educator         3.670692
engineer         3.541473
entertainment    3.440783
executive        3.349794
healthcare       2.896220
homemaker        3.301003
lawyer           3.735316
librarian        3.560781
marketing        3.485641
none             3.777778
other            3.552335
programmer       3.568205
retired          3.466750
salesman         3.582944
scientist        3.611273
student          3.515167
technician       3.532097
writer           3.376152
Name: rating, dtype: float64

**Observation:**: The mean rating is low (2.89) for people working in healthcare.

#### 10.2 Find mean rating for each gender

In [153]:
df_merge_all.groupby("gender")["rating"].mean()

gender
F    3.531510
M    3.529333
Name: rating, dtype: float64

**Observation**: The ratings given by men and women are similar on average.

#### 10.3 Find the mean rating grouped by both occupation and gender
1. Put both the columns under groupby function and find the mean rating. 

In [156]:
df_merge_all.groupby(["occupation", "gender"]).rating.mean().reset_index()

occupation gender    rating
0   administrator      F  3.781839
1   administrator      M  3.555233
2          artist      F  3.347065
3          artist      M  3.875841
4          doctor      M  3.688889
5        educator      F  3.699132
6        educator      M  3.660246
7        engineer      F  3.751724
8        engineer      M  3.537676
9   entertainment      F  3.448889
10  entertainment      M  3.439807
11      executive      F  3.773756
12      executive      M  3.320340
13     healthcare      F  2.736021
14     healthcare      M  3.639839
15      homemaker      F  3.278810
16      homemaker      M  3.500000
17         lawyer      F  3.623188
18         lawyer      M  3.741379
19      librarian      F  3.580070
20      librarian      M  3.537920
21      marketing      F  3.522624
22      marketing      M  3.474801
23           none      F  3.632877
24           none      M  3.876636
25          other      F  3.531114
26          other      M  3.563447
27     programmer      F  3.577566
28     programmer      M  3.567674
29        retired      F  3.239437
30        retired      M  3.477243
31       salesman      F  3.870206
32       salesman      M  3.394584
33      scientist      F  3.251799
34      scientist      M  3.637311
35        student      F  3.602879
36        student      M  3.484440
37     technician      F  3.268519
38     technician      M  3.540477
39         writer      F  3.663986
40         writer      M  3.180770

**Observations:** 

1. We had seen that the healthcare sector had a low mean rating. We now see that the low rating is driven by the female workers in the healthcare sector.

2. The male writers have given lower ratings compared to the female writers.

3. Only male doctors have given ratings. No female doctors have given ratings as per the data.

### Using groupby() on multiple columns of the data helps us analyze subsets of the data to gain insights.

### 11. Find the mean rating for each genre.

In [158]:
cols = df_merge_all.loc[:, "Action":"Western"].columns

for genre in cols:
    print(genre, ":", df_merge_all[df_merge_all[genre] == 1]["rating"].mean())

Action : 3.480245417953027
Adventure : 3.503526503308369
Animation : 3.5766990291262135
Childrens : 3.3532442216652742
Comedy : 3.3940734781442745
Crime : 3.6322780881440098
Documentary : 3.6728232189973613
Drama : 3.6873793708484772
Fantasy : 3.2152366863905324
Film-Noir : 3.9215233698788228
Horror : 3.2903893172841827
Musical : 3.521396851029471
Mystery : 3.63813155386082
Romance : 3.621704948358255
Sci-Fi : 3.5607227022780834
Thriller : 3.5090069495245064
War : 3.815811874866993
Western : 3.6132686084142396


**Observation**: War movies have the highest average rating, while Fantasy movies have the lowest average rating.

### 12. Divide the age column into four groups.

* Group 1. 0-20 => Teenager
* Group 2. 20-40 => Adult
* Group 3. 40-55 => Middle Age
* Group 4. 55+ => Elderly

This questions requires pandas cut() function.

Refer this link for more info: https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.cut.html

In [160]:
# create a new column with name age_group
df_merge_all["age_group"] = pd.cut(
    df_merge_all.age,
    bins=(0, 20, 40, 55, 100),
    labels=("Teenager", "Adult", "Middle Age", "Elderly"),
)

In [163]:
# print top 5 entries of the age and age_group columns
df_merge_all[["age", "age_group"]].head()

age age_group
0   60   Elderly
1   60   Elderly
2   60   Elderly
3   60   Elderly
4   60   Elderly

In [164]:
# print last 5 entries of the age and age_group columns
df_merge_all[["age", "age_group"]].tail()

age age_group
99985   19  Teenager
99986   19  Teenager
99987   19  Teenager
99988   19  Teenager
99989   19  Teenager

#### 12.1 Find the mean rating for each age group.

In [167]:
df_merge_all.groupby(["age_group"]).rating.mean()

age_group
Teenager      3.548290
Adult         3.492513
Middle Age    3.606980
Elderly       3.651703
Name: rating, dtype: float64

**Observation:** Old people have given higher ratings than any other age group.

#### 12.2 Find the mean, median, and standard deviation of rating for each age group.

1. This can be done using groupby.agg(), where agg() stands for aggregate.
2. agg() can take more than 1 aggregate function simultaneously.

Refer to this link for more info :https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html

In [169]:
df_merge_all.groupby(["age_group"])["rating"].agg(["mean", "median", "std"])

mean  median       std
age_group                             
Teenager    3.548290     4.0  1.154198
Adult       3.492513     4.0  1.138085
Middle Age  3.606980     4.0  1.083602
Elderly     3.651703     4.0  1.030762

#### 12.3 Which genre movies are common among Engineers?

In [172]:
df_merge_all.groupby("occupation").sum().loc[:, "Action":"Western"].loc[
    "doctor"
].sort_values(ascending=False)

Drama          238
Comedy         168
Romance        133
Thriller       117
Action         110
Adventure       57
Sci-Fi          56
War             44
Crime           41
Childrens       28
Musical         27
Mystery         27
Animation       22
Horror          18
Documentary     13
Film-Noir        9
Fantasy          7
Western          5
Name: doctor, dtype: int64

**Observation:**   Engineers mostly watch movies of Drama, Comedy, and Action genres.

You can try finding the most common genre among different professionals.

### We have just learned how to use the pandas cut() function to group a numerical column in the data into different categories and use agg() to use multiple aggregate functions at once with groupby(). These are important techniques that allow us to create informative groups from numerical columns and understand multiple statistics of subsets of our data at once.

In [173]:
df_merge_all.groupby("occupation").agg("mean")

movie id    Action  Adventure  Animation  Childrens  \
occupation                                                             
administrator  406.161653  0.248429   0.136382   0.027143   0.061773   
artist         429.569324  0.228769   0.134749   0.052426   0.087522   
doctor         438.129630  0.203704   0.105556   0.040741   0.051852   
educator       421.914522  0.207817   0.119479   0.027857   0.062811   
engineer       396.130781  0.278566   0.158919   0.044409   0.076951   
entertainment  442.510029  0.238300   0.118911   0.029608   0.049666   
executive      464.093180  0.237507   0.128160   0.026749   0.068195   
healthcare     568.310628  0.205777   0.115193   0.031740   0.075963   
homemaker      439.862876  0.307692   0.137124   0.030100   0.083612   
lawyer         422.749442  0.210409   0.105576   0.026022   0.061710   
librarian      428.190404  0.187370   0.101650   0.030154   0.054997   
marketing      396.340000  0.244615   0.126154   0.040513   0.083590   
none           459.886667  0.334444   0.177778   0.047778   0.101111   
other          420.551304  0.250610   0.134215   0.039486   0.082630   
programmer     412.141923  0.297692   0.158974   0.040000   0.074744   
retired        453.520199  0.172778   0.099441   0.021753   0.068365   
salesman       411.057243  0.259346   0.115654   0.017523   0.044393   
scientist      362.232264  0.275024   0.138484   0.029640   0.052478   
student        430.971944  0.291401   0.152942   0.041629   0.080115   
technician     386.343795  0.307846   0.159201   0.037660   0.063053   
writer         447.386992  0.233424   0.121771   0.031798   0.066305   

                 Comedy     Crime  Documentary     Drama   Fantasy  ...  \
occupation                                                          ...   
administrator  0.294558  0.075545     0.006418  0.414360  0.012569  ...   
artist         0.267331  0.087088     0.014731  0.414645  0.010399  ...   
doctor         0.311111  0.075926     0.024074  0.440741  0.012963  ...   
educator       0.286834  0.075204     0.011334  0.453448  0.011863  ...   
engineer       0.298263  0.074749     0.006239  0.385735  0.015292  ...   
entertainment  0.274117  0.098854     0.010984  0.383954  0.011939  ...   
executive      0.279541  0.084362     0.007643  0.413580  0.014697  ...   
healthcare     0.257489  0.071327     0.016049  0.462553  0.013552  ...   
homemaker      0.311037  0.083612     0.006689  0.347826  0.010033  ...   
lawyer         0.342007  0.079554     0.007435  0.403717  0.010409  ...   
librarian      0.298881  0.070169     0.009103  0.484923  0.010430  ...   
marketing      0.274359  0.082564     0.006154  0.422564  0.010256  ...   
none           0.283333  0.082222     0.001111  0.297778  0.013333  ...   
other          0.296849  0.077659     0.006096  0.403958  0.013506  ...   
programmer     0.310000  0.085385     0.006538  0.358974  0.015000  ...   
retired        0.318210  0.085146     0.009944  0.451833  0.013673  ...   
salesman       0.366822  0.092290     0.002336  0.370327  0.010514  ...   
scientist      0.259961  0.086006     0.008260  0.452867  0.009718  ...   
student        0.316907  0.083622     0.005329  0.354208  0.016032  ...   
technician     0.301284  0.078459     0.006847  0.369472  0.010842  ...   
writer         0.284372  0.089973     0.008311  0.398916  0.012827  ...   

                Mystery   Romance    Sci-Fi  Thriller       War   Western  \
occupation                                                                  
administrator  0.051477  0.205776  0.125953  0.209921  0.105094  0.022463   
artist         0.071057  0.200173  0.133449  0.206239  0.097487  0.012565   
doctor         0.050000  0.246296  0.103704  0.216667  0.081481  0.009259   
educator       0.056350  0.212477  0.104438  0.187162  0.106451  0.020019   
engineer       0.044409  0.179471  0.149376  0.209445  0.104967  0.022021   
entertainment  0.063037  0.163324  0.138013  0.264565  0.083095  0.012894   
executive      0.0